### ISSUE TITLE
xpu: implement grid_sample op for XPU (fallback to CPU not possible for fp16 and bf16)

### ISSUE THREAD

**Issue by dvrogozh** [pytorch/pytorch#127002]
grid_sample is currently not implemented for XPU backend, see
https://github.com/intel/torch-xpu-ops/blob/8cc6d5102878cefd3c71245fc7509a4548d07da2/src/aten/XPUFallback.template#L249

The problem is that grid_sample is also not implemented on CPU for fp16 and bf16 precisions, only for fp32 and fp64, see:
https://github.com/pytorch/pytorch/blob/c921c5cc779bbe80dfeb82db43ed61c8efaa3602/aten/src/ATen/native/GridSampler.cpp#L934-L935

Thus, runtime errors happen when running some networks which use grid_sample in fp16 or bf16. Examples are IFRNet and RIFE interpolation models. See:
* https://github.com/hzwer/ECCV2022-RIFE/blob/9a195b79f83a2e634f91369dfc9cb59cdd91fa84/model/warplayer.py#L22
* https://github.com/ltkong218/IFRNet/blob/b117bcafcf074b2de756b882f8a6ca02c3169bfe/utils.py#L21

This can be reproduced with:
```
import argparse
import torch
import sys

parser = argparse.ArgumentParser(prog='grid', description='grid')
parser.add_argument('--precision', action="store", type=str, required=False, default='int32')
args = parser.parse_args()

print('precision: ' + args.precision)
print('torch.__version__: ' + torch.__version__)
print('torch.xpu.is_available(): ' + str(torch.xpu.is_available()))
props = torch.xpu.get_device_properties()
print('xpu.has_fp16: ' + str(props.has_fp16))
print('xpu.has_fp64: ' + str(props.has_fp64))

if args.precision == 'bfloat16':
    input = torch.empty(1, 1, 2, 2, dtype=torch.bfloat16, device=torch.device('xpu'))
    grid = torch.empty(1, 1, 1, 2, dtype=torch.bfloat16, device=torch.device('xpu'))
elif args.precision == 'float16':
    input = torch.empty(1, 1, 2, 2, dtype=torch.float16, device=torch.device('xpu'))
    grid = torch.empty(1, 1, 1, 2, dtype=torch.float16, device=torch.device('xpu'))
elif args.precision == 'float32':
    input = torch.empty(1, 1, 2, 2, dtype=torch.float32, device=torch.device('xpu'))
    grid = torch.empty(1, 1, 1, 2, dtype=torch.float32, device=torch.device('xpu'))
elif args.precision == 'float64':
    input = torch.empty(1, 1, 2, 2, dtype=torch.float64, device=torch.device('xpu'))
    grid = torch.empty(1, 1, 1, 2, dtype=torch.float64, device=torch.device('xpu'))
else:
    print('unsupported precision: ' + args.precision)
    sys.exit(1)

res = torch.nn.functional.grid_sample(input, grid, align_corners=False)
print(res)
```

Output:
```
precision: float16
torch.__version__: 2.4.0a0+git7530cfe
torch.xpu.is_available(): True
xpu.has_fp16: True
xpu.has_fp64: True
/home/dvrogozh/git/pytorch/pytorch/torch/nn/functional.py:4368: UserWarning: The operator 'aten::grid_sampler_2d on the XPU backend is falling back to run on the CPU. (Triggered internally at /home/dvrogozh/git/pytorch/pytorch/build/aten/src/ATen/RegisterXPU.cpp:2920.)
  return torch.grid_sampler(input, grid, mode_enum, padding_mode_enum, align_corners)
Traceback (most recent call last):
  File "/home/dvrogozh/tests/grid.py", line 32, in <module>
    res = torch.nn.functional.grid_sample(input, grid, align_corners=False)
  File "/home/dvrogozh/git/pytorch/pytorch/torch/nn/functional.py", line 4368, in grid_sample
    return torch.grid_sampler(input, grid, mode_enum, padding_mode_enum, align_corners)
RuntimeError: grid_sampler_2d_cpu not implemented for Half
```

CC: @gujinghui @EikanWang @fengyuan14 @guangyey @jgong5

cc @gujinghui @EikanWang @fengyuan14 @guangyey

**Comment by EikanWang** [comment]
@fengyuan14 , let's implement grid_sampler_2d for XPU

**Comment by fengyuan14** [comment]
> @fengyuan14 , let's implement grid_sampler_2d for XPU

Sure, I will prioritize the operator.

**Comment by dvrogozh** [comment]
I see implementation is coming from @majing921201, see https://github.com/intel/torch-xpu-ops/pull/361.

**Comment by dvrogozh** [comment]
Done after merging https://github.com/pytorch/pytorch/pull/129353. Should be available in pytorch 2.5.
